project/ReflectiveCodeGen.scala

```diff
@@ -40,61 +42,72 @@ object ReflectiveCodeGen extends AutoPlugin {
-          try _underlying.run(request)
-          finally { _underlying = null }
-        else throw new IllegalStateException("Didn't set mutable generator")
+      watchSources ++= (watchSources in ProjectRef(file("."), "akka-grpc-codegen")).value,
+      watchSources ++= (watchSources in ProjectRef(file("."), "sbt-akka-grpc")).value,
+    )
 
-      override def suggestedDependencies: Seq[Artifact] =
-        if (_underlying ne null) _underlying.suggestedDependencies
-        else Nil
-
-      def setUnderlying(generator: protocbridge.ProtocCodeGenerator): Unit = _underlying = generator
-    }
-
-    val adapted = new MutableProtocCodeGenerator
-    MutableGeneratorAccess(adapted.setUnderlying _, (JvmGenerator("mutable", adapted), codeGeneratorSettings))
-  }
+  val setCodeGenerator = taskKey[Unit]("grpc-set-code-generator")
 
-  def loadAndSetGenerator(classpath: Classpath, mutableGeneratorClass: String, access: MutableGeneratorAccess): Unit = {
+  def loadAndSetGenerator(classpath: Classpath,
+                          languages: String,
+                          sources: String,
+                          extraGenerators: String,
+                          targetPath: File,
+                          generatorSettings: Seq[String],
+                          targets: ListBuffer[Target]): Unit = {
     val cp = classpath.map(_.data)
     // ensure to set right parent classloader, so that protocbridge.ProtocCodeGenerator etc are
     // compatible with what is already accessible from this sbt build
     val loader = ClasspathUtilities.toLoader(cp, classOf[protocbridge.ProtocCodeGenerator].getClassLoader)
-    val instance = loader.loadClass(mutableGeneratorClass).newInstance()
-    type WithInstance = {
-      def instance(): protocbridge.ProtocCodeGenerator
-    }
-    val generator = instance.asInstanceOf[WithInstance].instance
-    access.setUnderlying(generator)
+    import scala.reflect.runtime.universe
+    import scala.tools.reflect.ToolBox
+
+    val tb = universe.runtimeMirror(loader).mkToolBox()
+    val source =
+      s"""import akka.grpc.sbt.AkkaGrpcPlugin
+          |import AkkaGrpcPlugin.autoImport._
+          |import AkkaGrpc._
+          |import akka.grpc.gen.scaladsl._
+          |import akka.grpc.gen.javadsl._
+          |
+          |val languages: Seq[AkkaGrpc.Language] = Seq($languages)
+          |val sources: Seq[AkkaGrpc.GeneratedSource] = Seq($sources)
+          |
+          |val logger = akka.grpc.gen.StdoutLogger
+          |
+          |(targetPath: java.io.File, settings: Seq[String]) => {
+          |  val generators =
+          |    AkkaGrpcPlugin.generatorsFor(sources, languages, logger) ++
+          |    Seq($extraGenerators).map(gen => AkkaGrpcPlugin.toGenerator(gen, akka.grpc.gen.StdoutLogger))
+          |  AkkaGrpcPlugin.targetsFor(targetPath, settings, generators)
+          |}
+        """.stripMargin
+    val generatorsF = tb.eval(tb.parse(source)).asInstanceOf[(File, Seq[String]) => Seq[Target]]
```

Comment:
:open_mouth: cool

Reply:
lol that's some old school stuff there lol, have not seen this style of reflection since ages :)